src/Dofs/sparsity_pattern.jl

```diff
@@ -1,83 +1,473 @@
+    write(io, seekstart(iob))
+    return
+end
+
+n_rows(sp::SparsityPattern) = sp.nrows
+n_cols(sp::SparsityPattern) = sp.ncols
+
+@inline function add_entry!(sp::SparsityPattern, row::Int, col::Int)
+    @boundscheck (1 <= row <= n_rows(sp) && 1 <= col <= n_cols(sp)) || throw(BoundsError(sp, (row, col)))
+    r = @inbounds sp.rows[row]
+    r = insert_sorted(r, col)
+    @inbounds sp.rows[row] = r
+    return
+end
+
+@inline function insert_sorted(x::HeapAllocator.HeapVector{Int}, item::Int)
+    k = searchsortedfirst(x, item)
+    if k == length(x) + 1 || @inbounds(x[k]) != item
+        x = HeapAllocator.insert(x, k, item)
+    end
+    return x
+end
+
+eachrow(sp::SparsityPattern)           = sp.rows
+eachrow(sp::SparsityPattern, row::Int) = sp.rows[row]
+
+
+################################################
+## Adding entries to AbstractSparsityPatterns ##
+################################################
+
+"""
+    init_sparsity_pattern(dh::DofHandler; nnz_per_row::Int)
+
+Initialize an empty [`SparsityPattern`](@ref) with `ndofs(dh)` rows and `ndofs(dh)` columns.
+
+# Keyword arguments
+ - `nnz_per_row`: memory optimization hint for the number of non-zero entries per row that
+   will be added to the pattern.
+"""
+function init_sparsity_pattern(
+        dh::DofHandler;
+        # TODO: What is a good estimate for nnz_per_row?
+        nnz_per_row::Int = 2 * ndofs_per_cell(dh.subdofhandlers[1]), # FIXME
+    )
+    sp = SparsityPattern(ndofs(dh), ndofs(dh); nnz_per_row = nnz_per_row)
+    return sp
+end
+
+"""
+    create_sparsity_pattern!(
+        sp::AbstractSparsityPattern,
+        dh::DofHandler,
+        ch::Union{ConstraintHandler, Nothing} = nothing;
+        topology = nothing,
+        keep_constrained::Bool = true,
+        coupling = nothing,
+        interface_coupling = nothing,
+    )
+
+Convenience method for doing the common task of calling [`add_cell_entries!`](@ref),
+[`add_interface_entries!`](@ref), and [`add_constraint_entries!`](@ref), depending on what
+arguments are passed:
+ - `add_cell_entries!` is always called
+ - `add_interface_entries!` is called if `topology` is provided (i.e. not `nothing`)
+ - `add_constraint_entries!` is called if the ConstraintHandler is provided
+
+For more details about arguments and keyword arguments, see the respective functions.
 """
-    create_sparsity_pattern(dh::DofHandler; coupling, [topology::Union{Nothing, AbstractTopology}], [cross_coupling])
+function create_sparsity_pattern!(
+        sp::AbstractSparsityPattern, dh::DofHandler,
+        ch::Union{ConstraintHandler, Nothing} = nothing;
+        keep_constrained::Bool = true,
+        coupling::Union{AbstractMatrix{Bool}, Nothing} = nothing,
+        interface_coupling::Union{AbstractMatrix{Bool}, Nothing} = nothing,
+        topology = nothing,
+    )
+    # Argument checking
+    isclosed(dh) || error("the DofHandler must be closed")
+    if n_rows(sp) < ndofs(dh) || n_cols(sp) < ndofs(dh)
```

Comment:
Out of curiosity, why is this `<` and not `!=`?

Reply:
You can create a pattern with ndofs+x rows/cols and have entries from 1:ndofs inserted in the first block.